### PR TITLE
Scene generation feature

### PR DESCRIPTION
**Pull request by pawel-kotowski**
This PR is ready for code review and testing.  To test it enable [WarehouseAssets](https://github.com/RobotecAI/o3de-extras/tree/development/Gems/WarehouseAssets) gem. 

[Readme.md](https://github.com/RobotecAI/o3de-genai-gems/blob/pk/scene_generation_feature/Framework/GenAIFramework/Editor/Scripts/scene_generation/README.md)

### CODE REVIEW

File: Framework/GenAIFramework/Code/Source/O3DESceneGenerationAssistantPythonFeature/O3DESceneGenerationAssistantPythonFeature.h [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+#include <GenAIFramework/Feature/FeatureBase.h>
+#include <GenAIFramework/Feature/PythonFeatureBase.h>
+#include <GenAIFramework/GenAIFrameworkTypeIds.h>
+
+#include <AzCore/RTTI/RTTI.h>
+#include <AzCore/Serialization/SerializeContext.h>
+
+namespace GenAIFramework
+{
+    class O3DESceneGenerationAssistantPythonFeature : public PythonFeatureBase
```
Comment (jhanca-robotecai) [inline Framework/GenAIFramework/Code/Source/O3DESceneGenerationAssistantPythonFeature/O3DESceneGenerationAssistantPythonFeature.h:20]: I would keep the names shorter: `SceneGenerationPythonFeature` is more than enough.

File: Framework/GenAIFramework/Code/Source/O3DESceneGenerationAssistantPythonFeature/O3DESceneGenerationAssistantPythonFeature.h [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    {
+    public:
+        AZ_RTTI(O3DESceneGenerationAssistantPythonFeature, "{7c524a1b-a175-4c26-beb1-dbefb7f13f50}", PythonFeatureBase);
```
Comment (jhanca-robotecai) [inline Framework/GenAIFramework/Code/Source/O3DESceneGenerationAssistantPythonFeature/O3DESceneGenerationAssistantPythonFeature.h:23]: Please move _TypeId_ to a dedicated file, which is even included.

File: Framework/GenAIFramework/Code/Source/O3DESceneGenerationAssistantPythonFeature/O3DESceneGenerationAssistantPythonFeature.h [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+        O3DESceneGenerationAssistantPythonFeature() = default;
+        O3DESceneGenerationAssistantPythonFeature(AZ::u64 agentId, AZ::u64 conversationId);
+        ~O3DESceneGenerationAssistantPythonFeature() = default;
+
+        static void Reflect(AZ::ReflectContext* context);
+
+    private:
+        constexpr static const char* PythonSceneGenerationAssistantScript = "Editor/Scripts/scene_generation/scene_generation_assistant.py";
+
+        AZStd::string m_sceneGenerationAssistantPythonScriptLocation;
```
Comment (jhanca-robotecai) [inline Framework/GenAIFramework/Code/Source/O3DESceneGenerationAssistantPythonFeature/O3DESceneGenerationAssistantPythonFeature.h:35]: Please keep the names shorter: this is a member of a class called _SceneGenerationAssistant_, no need to repeat it.

File: Framework/GenAIFramework/Editor/Scripts/scene_generation/README.md [hunk 4]
```diff
@@ -0,0 +1,27 @@
+# Scene generation feature
+
+This feature demonstrates the use of the GenAIFramework for scene generation. Language model is used for Python code generation that manipulates prefabs through the interface available in [Python Editor Bindings](https://docs.o3de.org/docs/user-guide/editor/editor-automation/).
```
Comment (jhanca-robotecai) [inline Framework/GenAIFramework/Editor/Scripts/scene_generation/README.md:3]: `The language model`

File: Framework/GenAIFramework/Code/Source/O3DESceneGenerationAssistantPythonFeature/O3DESceneGenerationAssistantPythonFeature.h (hunk 3 above)
Comment (pawel-kotowski) [reply Framework/GenAIFramework/Code/Source/O3DESceneGenerationAssistantPythonFeature/O3DESceneGenerationAssistantPythonFeature.h:35]: I was following the convention from o3de assistant example (and went too far, but if we simplify it here then let's also simplify it there)

File: Framework/GenAIFramework/Editor/Scripts/scene_generation/README.md (hunk 4 above)
Comment (pawel-kotowski) [reply Framework/GenAIFramework/Editor/Scripts/scene_generation/README.md:3]: I think we don't refer to a specific model so no "the" is ok.

File: Framework/GenAIFramework/Code/Source/O3DESceneGenerationAssistantPythonFeature/O3DESceneGenerationAssistantPythonFeature.h (hunk 3 above)
Comment (pawel-kotowski) [reply Framework/GenAIFramework/Code/Source/O3DESceneGenerationAssistantPythonFeature/O3DESceneGenerationAssistantPythonFeature.h]: Applied [8976295](https://github.com/RobotecAI/o3de-genai-gems/pull/158/commits/89762950bc0059d16d7fc100f1f8ee4e8275401c). Also names in O3DE Assistant (Python) have been adapted.

File: Framework/GenAIFramework/Code/Source/O3DESceneGenerationAssistantPythonFeature/O3DESceneGenerationAssistantPythonFeature.h (hunk 2 above)
Comment (pawel-kotowski) [reply Framework/GenAIFramework/Code/Source/O3DESceneGenerationAssistantPythonFeature/O3DESceneGenerationAssistantPythonFeature.h]: Applied [8976295](https://github.com/RobotecAI/o3de-genai-gems/pull/158/commits/89762950bc0059d16d7fc100f1f8ee4e8275401c)

File: Framework/GenAIFramework/Code/Source/O3DESceneGenerationAssistantPythonFeature/O3DESceneGenerationAssistantPythonFeature.h (hunk 1 above)
Comment (pawel-kotowski) [reply Framework/GenAIFramework/Code/Source/O3DESceneGenerationAssistantPythonFeature/O3DESceneGenerationAssistantPythonFeature.h]: Applied [8976295](https://github.com/RobotecAI/o3de-genai-gems/pull/158/commits/89762950bc0059d16d7fc100f1f8ee4e8275401c)

File: Framework/GenAIFramework/Editor/Scripts/scene_generation/README.md (hunk 4 above)
Comment (pawel-kotowski) [reply Framework/GenAIFramework/Editor/Scripts/scene_generation/README.md]: I have moved the readme to gettingStarted section [42d7eb5](https://github.com/RobotecAI/o3de-genai-gems/pull/158/commits/42d7eb54204beeea9d40058c5ae2bb83f12a1a0b)

File: Framework/GenAIFramework/Editor/Scripts/scene_generation/prefab.py [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+        return self.name == other.name
+
+
+class SpawnedPrefab:
+    def __init__(
+        self,
+        entity_id: EditorEntityId,
+        translation: list[float],
+        available_prefab: AvailablePrefab,
+        overlapping_entity_ids: set[int] = None,
```
Comment (rachwalk) [inline Framework/GenAIFramework/Editor/Scripts/scene_generation/prefab.py:137]:
```suggestion
        overlapping_entity_ids: set[int] | None = None,
```

File: Framework/GenAIFramework/Editor/Scripts/scene_generation/prefab.py [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+        name: str,
+        path: str,
+        description: str,
+        size: list[float],
+        outer_margin: list[float] = None,
+        anchor_transform: dict[str, list[float]] = None,
+        excluded_from_overlap_check: bool = False,
+        can_be_put_on: list[str] = None,
+        shelves_zs: list[float] = None,
+        accesibility_requirements: str = None,
```
Comment (rachwalk) [inline Framework/GenAIFramework/Editor/Scripts/scene_generation/prefab.py:58]:
```suggestion
        outer_margin: list[float] | None = None,
        anchor_transform: dict[str, list[float]] | None = None,
        excluded_from_overlap_check: bool = False,
        can_be_put_on: list[str] | None = None,
        shelves_zs: list[float] | None = None,
        accesibility_requirements: str | None = None,
```

File: Framework/GenAIFramework/Editor/Scripts/scene_generation/prefabs_manager.py [hunk 7]
```diff
@@ -0,0 +1,421 @@
...
+
+
+class PrefabsManager:
+    def __init__(
+        self,
+        prefabs_info_yaml_path: None | Path = None,
+    ) -> None:
+        self.prefabs_info_yaml_path = prefabs_info_yaml_path
+        self.automatically_removed_prefabs = set()
+        self.actions_log = []
```
Comment (rachwalk) [inline Framework/GenAIFramework/Editor/Scripts/scene_generation/prefabs_manager.py:43]: Add type annotations
```suggestion
        self.automatically_removed_prefabs : set[SpawnedPrefab] = set()
        self.actions_log : list[str] = []
```

File: Framework/GenAIFramework/Editor/Scripts/scene_generation/prefabs_manager.py [hunk 8]
```diff
@@ -0,0 +1,421 @@
...
+        self.available_prefabs = {}
```
Comment (rachwalk) [inline Framework/GenAIFramework/Editor/Scripts/scene_generation/prefabs_manager.py:44]: Why not initialize it immediately? Also provides automatic type inference.
```suggestion
        self.available_prefabs = self.get_available_prefabs()
```

File: Framework/GenAIFramework/Editor/Scripts/scene_generation/scene_generation.py [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+PREFABS_YAML_FILE_PATH = Path(__file__).parent / "prefabs.yml"
+
+
+class SceneGeneration:
+    def __init__(self, agent_id):
+        self.agent_id = agent_id
+        self.pm = PrefabsManager.from_yaml(PREFABS_YAML_FILE_PATH)
+        self.messages: list[GenAIFrameworkMessage] = []
+        self.python_output: str = ""
+        self.response_with_status: dict[str, bool | GenAIFrameworkMessage] = None
```
Comment (rachwalk) [inline Framework/GenAIFramework/Editor/Scripts/scene_generation/scene_generation.py:56]:
```suggestion
        self.response_with_status: dict[str, bool | GenAIFrameworkMessage] | None= None
```

File: Framework/GenAIFramework/Editor/Scripts/scene_generation/utils.py [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+        return [
+            round(float(match.group(1)), 2),
+            round(float(match.group(2)), 2),
+            round(float(match.group(3)), 2),
+        ]
+    else:
+        raise ValueError(f"Invalid vector string: {vector_str}")
+
+
+def extract_text_from_xml_tags(text: str, tag: str = "python") -> str:
```
Comment (rachwalk) [inline Framework/GenAIFramework/Editor/Scripts/scene_generation/utils.py:60]:
```suggestion
def extract_text_from_xml_tags(text: str, tag: str = "python") -> str  | None:
```

**Review by rachwalk — CHANGES_REQUESTED** [review]
I've made a few suggestions regarding python typing

File: Framework/GenAIFramework/Editor/Scripts/scene_generation/prefab.py [hunk 11]
```diff
@@ -0,0 +1,260 @@
...
+    ) -> None:
+        self.name = name
+        self.path = path
+        self.description = description
+        self.size = size
+        self.outer_margin = outer_margin
+        self.shelves_zs = shelves_zs
+        self.can_be_put_on = can_be_put_on
+        self.anchor_transform = anchor_transform or {
```
[the remark this answers is not in the document]
Comment (pawel-kotowski) [reply Framework/GenAIFramework/Editor/Scripts/scene_generation/prefab.py:67]: Mutable default arguments are shared between instances, so it's not a good idea in general.

File: Framework/GenAIFramework/Editor/Scripts/scene_generation/prefabs_manager.py (hunk 8 above)
Comment (pawel-kotowski) [reply Framework/GenAIFramework/Editor/Scripts/scene_generation/prefabs_manager.py:44]: We read available prefabs from yaml on each request anyway to allow modyfing prefabs without recreating the chat window.

File: Framework/GenAIFramework/Editor/Scripts/scene_generation/prefab.py [hunk 12]
```diff
@@ -0,0 +1,260 @@
...
+from typing import Any
+from utils import prefab_dict_to_xml
+import azlmbr
+
+
+def get_overlapping_prefabs_pairs(
+    prefabs: set["SpawnedPrefab"],
+) -> list[tuple["SpawnedPrefab", "SpawnedPrefab"]]:
+    """Gets pairs of overlapping prefabs. Ignores prefabs with different z coord."""
+    prefabs = list(prefabs)
```
[the remark this answers is not in the document]
Comment (pawel-kotowski) [reply Framework/GenAIFramework/Editor/Scripts/scene_generation/prefab.py]: Applied [fb22647](https://github.com/RobotecAI/o3de-genai-gems/pull/158/commits/fb22647b007c73952202ad290df598886b1b52e0)

File: Framework/GenAIFramework/Editor/Scripts/scene_generation/prefab.py (hunk 5 above)
Comment (pawel-kotowski) [reply Framework/GenAIFramework/Editor/Scripts/scene_generation/prefab.py]: Applied [fd7b2ce](https://github.com/RobotecAI/o3de-genai-gems/pull/158/commits/fd7b2cecb8dceff013efdf449ecaa331c85f82a6)

File: Framework/GenAIFramework/Editor/Scripts/scene_generation/prefab.py (hunk 6 above)
Comment (pawel-kotowski) [reply Framework/GenAIFramework/Editor/Scripts/scene_generation/prefab.py]: Applied [fd7b2ce](https://github.com/RobotecAI/o3de-genai-gems/pull/158/commits/fd7b2cecb8dceff013efdf449ecaa331c85f82a6)

File: Framework/GenAIFramework/Editor/Scripts/scene_generation/scene_generation.py (hunk 9 above)
Comment (pawel-kotowski) [reply Framework/GenAIFramework/Editor/Scripts/scene_generation/scene_generation.py]: Applied [fd7b2ce](https://github.com/RobotecAI/o3de-genai-gems/pull/158/commits/fd7b2cecb8dceff013efdf449ecaa331c85f82a6)

File: Framework/GenAIFramework/Editor/Scripts/scene_generation/utils.py (hunk 10 above)
Comment (pawel-kotowski) [reply Framework/GenAIFramework/Editor/Scripts/scene_generation/utils.py]: Applied [fd7b2ce](https://github.com/RobotecAI/o3de-genai-gems/pull/158/commits/fd7b2cecb8dceff013efdf449ecaa331c85f82a6)

File: Framework/GenAIFramework/Editor/Scripts/scene_generation/prefabs_manager.py (hunk 7 above)
Comment (pawel-kotowski) [reply Framework/GenAIFramework/Editor/Scripts/scene_generation/prefabs_manager.py:43]: Applied [fd7b2ce](https://github.com/RobotecAI/o3de-genai-gems/pull/158/commits/fd7b2cecb8dceff013efdf449ecaa331c85f82a6)

**Comment by pawel-kotowski** [comment]
@jhanca-robotecai I have also changed the names of o3de assistant Python file and class, from `assistant` to `o3de_assistant`. The scene generation Python file is now called simply `scene_generation`. This way it's more consistent with names in C++ code.

**Comment by pawel-kotowski** [comment]
I have added showing error messages in the chat to the user [af439d1](https://github.com/RobotecAI/o3de-genai-gems/pull/158/commits/af439d14fdc2f751698652ca5747bddfe7b45f46)

**Comment by pawel-kotowski** [comment]
@jhanca-robotecai @rachwalk Thank you for you reviews, I think I've applied all suggestions. Please check if the result is ok.

**Review by rachwalk — APPROVED** [review]
lgtm

**Review by jhanca-robotecai — APPROVED** [review]
Some of my comments are still in place, but I will refer to them in new issues.

The code sometimes works correctly, but often fails with Error 400 or Error 500. I will approve it as a work in progress pushed to the development.

**Pull request merged**